watchtower/wtdb/client_db.go

```diff
@@ -1001,6 +1017,120 @@ func (c *ClientDB) FetchSessionCommittedUpdates(id *SessionID) (
 	return committedUpdates, nil
 }
 
+// IsAcked returns true if the given backup has been backed up using the given
+// session.
+func (c *ClientDB) IsAcked(id *SessionID, backupID *BackupID) (bool, error) {
+	var isAcked bool
+	err := kvdb.View(c.db, func(tx kvdb.RTx) error {
+		sessions := tx.ReadBucket(cSessionBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		chanDetailsBkt := tx.ReadBucket(cChanDetailsBkt)
+		if chanDetailsBkt == nil {
+			return ErrUninitializedDB
+		}
+
+		sessionBkt := sessions.NestedReadBucket(id[:])
+		if sessionsBkt == nil {
+			return nil
+		}
+
+		// Get the DB representation of the channel-ID.
+		dbChanID, _, err := getDBChanID(chanDetailsBkt, backupID.ChanID)
+		if err != nil {
+			return err
+		}
+
+		// Get the DB representation of the session-ID.
+		dbSessID, dbSessIDBytes, err := getDBSessionID(sessions, *id)
+		if err != nil {
+			return err
+		}
+
+		sessionAckRanges := sessionBkt.NestedReadBucket(
+			cSessionAckRangeIndex,
+		)
+		if sessionAckRanges == nil {
+			return nil
+		}
+
+		rangesBkt := sessionAckRanges.NestedReadBucket(dbSessIDBytes)
+		if rangesBkt == nil {
+			return nil
+		}
+
+		index, err := c.getRangeIndex(rangesBkt, dbSessID, dbChanID)
+		if err != nil {
+			return err
+		}
+
+		isAcked = index.IsInIndex(backupID.CommitHeight)
+		return nil
+	}, func() {})
+	if err != nil {
+		return false, err
+	}
+
+	return isAcked, nil
+}
+
+// NumAckedUpdates returns the number of backups that have been successfully
+// backed up using the given session.
+func (c *ClientDB) NumAckedUpdates(id *SessionID) (uint64, error) {
+	var numAcked uint64
+	err := kvdb.View(c.db, func(tx kvdb.RTx) error {
+		sessions := tx.ReadBucket(cSessionBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		sessionBkt := sessions.NestedReadBucket(id[:])
+		if sessionsBkt == nil {
+			return nil
+		}
+
+		sessionAckRanges := sessionBkt.NestedReadBucket(
+			cSessionAckRangeIndex,
+		)
+		if sessionAckRanges == nil {
+			return nil
+		}
+
+		dbSessionID, _, err := getDBSessionID(sessions, *id)
+		if err != nil {
+			return err
+		}
+
+		// Iterate over the channel ID's in the sessionAckRanges
+		// bucket.
+		return sessionAckRanges.ForEach(func(dbChanID, _ []byte) error {
+			rangesBkt := sessionAckRanges.NestedReadBucket(dbChanID)
+			if rangesBkt == nil {
+				return nil
+			}
+
+			// Get the range index for the session-channel pair.
+			chanID := byteOrder.Uint64(dbChanID)
+			index, err := c.getRangeIndex(
```

Comment:
Here we could also use the in-memory cache exclusively if we know that all range indexes for all channels are already cached. Maybe not as critical if `NumAckedUpdates` isn't called too frequently.

Reply:
yeah the reason I didnt use the in-mem one here is cause this iterates through all sessions (even ones we will never use again) and would result in them then being stored in the cache even if they are not ever needed again